pkg/v1/sdk/capabilities/discovery/cluster_gvr.go

```diff
@@ -35,94 +58,239 @@ func (q *QueryGVR) Name() string {
+	var unmatched []string
+	for _, ver := range q.versions {
+		gvr := schema.GroupVersionResource{
+			Group:    q.group,
+			Version:  ver,
+			Resource: q.resource.String,
+		}
+		resources, err := cfg.discoveryClientset.ServerResourcesForGroupVersion(gvr.GroupVersion().String())
 		if err != nil {
 			// Second condition is because fake discovery client does not return a proper NotFound error.
-			if apierrors.IsNotFound(err) || strings.Contains(err.Error(), fmt.Sprintf("GroupVersion %q not found", gvr.GroupVersion().String())) {
+			if apierrors.IsNotFound(err) || strings.Contains(
+				err.Error(),
+				fmt.Sprintf("GroupVersion %q not found", gvr.GroupVersion().String()),
+			) {
 				unmatched = append(unmatched, gvr.String())
-				continue
+			} else {
+				return nil, err
 			}
-			return false, err
-		}
-
-		// Check resource if specified.
-		if gvr.Resource == "" {
-			continue
 		}
 		if !q.resourceExists(resources) {
 			unmatched = append(unmatched, gvr.String())
-			continue
 		}
 	}
-
-	q.unmatchedGVRs = unmatched
-	return len(unmatched) == 0, nil
+	return unmatched, nil
 }
 
-func (q *QueryGVR) groupExists(group string, groupList *metav1.APIGroupList) bool {
-	for i := range groupList.Groups {
-		if strings.EqualFold(groupList.Groups[i].Name, group) {
+// groupsContains looks for a group in a slice of APIGroup.
+func (q *QueryGVR) groupsContains(groups []*metav1.APIGroup, group string) bool {
+	for _, grp := range groups {
+		if strings.EqualFold(grp.Name, group) {
 			return true
 		}
 	}
 	return false
 }
 
+// resourceListInGroup looks within a slice of APIResourceList for an
+// APIResourceList that is in a specified group.
+func (q *QueryGVR) resourceListInGroup(resourceLists []*metav1.APIResourceList, group string) *metav1.APIResourceList {
+	for _, rscList := range resourceLists {
+		if strings.EqualFold(extractGroup(rscList.GroupVersion), group) {
+			return rscList
+		}
+	}
+	return nil
+}
+
+// groupFromGroupList looks for a particular group from an APIGroupList.
+func (q *QueryGVR) groupFromGroupList(groupList *metav1.APIGroupList) *metav1.APIGroup {
+	for i := range groupList.Groups {
+		if strings.EqualFold(groupList.Groups[i].Name, q.group) {
+			return &groupList.Groups[i]
+		}
+	}
+	return nil
+}
+
 func (q *QueryGVR) resourceExists(resources *metav1.APIResourceList) bool {
+	if resources == nil {
+		return false
+	}
+
 	for i := range resources.APIResources {
-		if resources.APIResources[i].Name == q.resource {
+		if resources.APIResources[i].Name == q.resource.String {
 			return true
 		}
 	}
 	return false
 }
 
-// Reason surfaces what didnt match
+// extractGroup extracts a group from the GroupVersion field In APIResourceList.
+// Examples are,
+// - GroupVersion:"autoscaling/v1"
+// - GroupVersion:"v1"
+// In the first example, the group returned will be "autoscaling".
+// In the second example, the group returned will be an empty string.
+func extractGroup(groupVersion string) string {
```

Comment:
optional: we could use [this function](https://github.com/kubernetes/apimachinery/blob/57893b822e174aa3bdaeee434bc42d22b37fc2dd/pkg/runtime/schema/group_version.go#L209-L211) to extract group to be future-proof. But if you think it's an overkill, I'm okay with this code.

Reply:
Oh, good find. I like using existing functions provided by Kubernetes.

Reply:
I got rid of the `extractGroup` function and used Kubernetes' `schema.ParseGroupVersion(...)` instead.